### ISSUE TITLE
One of the futures returned by an async API gets stuck forever when auto flush is disabled

### ISSUE THREAD

**Issue by dbarvitsky** [lettuce-io/lettuce-core#2345]
## Bug Report

The `connection.setAutoFlushCommands(false)` and `connection.flushCommands()` cause *one* (!) out of many futures returned by async commands to never (?) resolve, while all commands are executed by Redis successfully.

There are several similar issues concerned infinite (probably) waits on futures returned by async():
- https://github.com/lettuce-io/lettuce-core/issues/2261
- https://github.com/lettuce-io/lettuce-core/issues/542

#### Current Behavior

<details>
<summary>Stack trace</summary>

No stack trace, code doesn't complete in a reasonable time.

</details>

#### Input Code

Please see the full unit test in details. The test creates arbitrary large number of asynchronous commands and polls for completion, while printing counters (updated by futures' listeners). When the number is small (i.e. 100K) everything seems to be working OK. When a number is large (500K for me), all futures complete except for one, that is stuck indefinitely.

Note that all commands actually execute in Redis (`hlen test` gives the expected count, it is just 1 future that doesn't get resolved for some reason).

Playing with the client resources configuration didn't change the behavior. However `connection.setAutoFlushCommands(true)` seemed to fix the issue (although test is running ~5-7 times slower). 

The test is designed to reproduce a real problem we see under load periodically, when cache seed jobs all of sudden begin to wait excessively after each batch (10K).

<details>
<summary>Input Code</summary>

```java
 @Test
  public void testAsyncCommands() throws InterruptedException {
    // The scenario is sensitive to this number, 500K reproduces it 100% for me, something like
    // 163K does 7 out of 10.
    final var bigNumber = 500_000;
    // Default client resources:
    final var resources = DefaultClientResources.builder().build();
    try {
      final var redis = RedisClient.create(resources, this.onlineStoreUri);
      final var connection = redis.connect();
      // Manually flush commands in batch mode, as suggested by wiki.
      connection.setAutoFlushCommands(false);
      // Time out the connection aggressively. I would expect each command not to run longer than
      // 1 second with this setting.
      connection.setTimeout(1000, TimeUnit.MILLISECONDS);  
      // The list of futures we will be monitoring, plus some additional counters we will be calling
      // from the futures' handlers.    
      final var futures = new ArrayList<Future<?>>();      
      final var errorCounter = new AtomicInteger(0);
      final var successCounter = new AtomicInteger(0);
      final var remainingCounter = new AtomicInteger(0);
      try {
        final var commands = connection.async();
        for (int i = 0; i < bigNumber; i++) {
          final var k = String.valueOf(i);
          remainingCounter.incrementAndGet();
          final var future = commands.hmset("test", Map.of(k,k));
          futures.add(future);
          // Add a a handler to see when commands complete.
          future.whenComplete((response, error) -> {
              if (error != null) {
                errorCounter.incrementAndGet();
              } else {
                successCounter.incrementAndGet();
              }
              remainingCounter.decrementAndGet();
            });
        }
        // Flush commands as per wiki.
        connection.flushCommands();

        // waitAll(futures) just hangs, so we are iterating over them instead and printing counters
        // every second.
        while (remainingCounter.get() > 0) {
          System.out.println("%d remaining, %d succeeded, %d failed, %d futures remaining"
              .formatted(
                remainingCounter.get(),
                successCounter.get(),
                errorCounter.get(),
                futures.stream().filter(f -> !f.isDone()).count()
              )
          );
          Thread.sleep(1000);       
        }
      } finally {
        connection.close();
      }
    } finally {
      resources.shutdown(10000, 20000, TimeUnit.MILLISECONDS);
    }
  }
```

</details>

#### Expected behavior/code

I would expect the returned futures to either succeed or fail as soon as Redis commands have been carried out. I would prefer async APIs to fail fast if the async operation cannot be scheduled.

#### Environment

- Lettuce version(s): 4.5.0.Final
- Redis version: 7.0.7 (AWS ElastiCache)

#### Additional context

This is used for batch-seeding the cache with key space ~ 250M. While it is sustainable to do so with `setAutoFlushCommands(true)`, the load times are significantly higher.

**Comment by dbarvitsky** [comment]
Thank you in advance for your time and consideration!

**Comment by mp911de** [comment]
Thanks for reaching out. Manual flushing is to improve flushing behavior. If you manage to combine commands into blocks of 8-16kB of data, then you typically yield the best ratio between overhead and throughput. Please take into account, the more commands you buffer on the client side, the more JVM memory you require. That can become inefficient at some point.

I therefore suggest that you perform some sort of batching (e.g. flushing after 10 to 50 commands) instead of accumulating 100.000 command objects. First, this makes it hard to diagnose which command has failed and secondly, reading logs becomes impossible to trace down the actual cause.

**Comment by dbarvitsky** [comment]
@mp911de thank you for suggestions.

The ~50 command batches do work, and the off-by-one does not seem to happen (tested in an isolated experiment). The small batches (under 1K) do not perform well in our specific case (we tuned the 10K-per-batch number to our setup and workload experimentally, so I don't have a sound explanation why exactly).

Your assertions about the dangers of large batches certainly are valid, and I would too advise against just throwing in a big number without experimentation. I can elaborate on why we decided to take the risk with 10K batches if you are interested. "There are no solutions in cloud computing, just trade-offs" ;-)

The good news is that upgrade to `6.2.4.RELEASE` (which we finally were able to do) seems to be completely eliminating the issue regardless of the batch size. At least the code above no longer hangs. I can't say what exactly fixed it between 4.5.0 and 6.2.4, but the upgrade seemed to have solved the problem for us.

Thank you!
